### PR TITLE
Client side errors

### PR DESCRIPTION
**Pull request by abhinav**
This implements support for consistent errors on the client side. Clients will
only see `transport.UnexpectedError` or `transport.BadRequestError`.

The design for errors is:

- Certain error types are `HandlerError`s. The Inbound has to know how to
  handle these. Examples: BadRequest and Unexpected.
- Other error types know how to convert themselves into HandlerErrors by
  implementing proving a `AsHandlerError() error` function. For example,
  UnknownProcedureError knows that it should be converted into a
  BadRequestError.
- All other error types are converted into UnexpectedErrors with information
  about the procedure and service.

The hierarchy we have is,

```
BadRequestError                 HandlerError
 |                                        |
 +--------> localBadRequestError <--------+
 |                                        |
 +--------> remoteBadRequestError         |
                                          |
UnexpectedError                           |
 |                                        |
 +--------> localUnexpectedError <--------+
 |
 +--------> remoteUnexpectedError
```

Note that only the `local` versions of the error types are valid
`HandlerError`s. That is, only the local versions make it to the Inbound as-is
and must be handled (resulting in a 400-level error for `BadRequest` and
500-level for `Unexpected`). If a remote version of either error is returned by
a `Handler`, it causes a *new* `UnexpectedError` which includes the procedure
name and service for that call. This way, we end up with full information for
multi-hop failures:

    UnexpectedError: error for procedure "foo" of service "a":
        UnexpectedError: error for procedure "bar" of service "b":
            BadRequest: unrecognized procedure "baz" for service "c"

----

Note that this is a PR against #106.

### CODE REVIEW

**Comment by abhinav** [comment]
@yarpc/yarpc

**Comment by breerly** [comment]
@abhinav wondering if you can update `errors.md` to make it reflect the reality we want

**Comment by breerly** [comment]
```
UnexpectedError: error for procedure "foo" of service "a":
    UnexpectedError: error for procedure "bar" of service "b":
        BadRequest: unrecognized procedure "baz" for service "c"
```

👏

File: transport/http/outbound.go [hunk 1]
```diff
@@ -76,15 +76,24 @@ func (o outbound) Call(ctx context.Context, req *transport.Request) (*transport.
...
 		if err := response.Body.Close(); err != nil {
-			return nil, err // TODO error type
+			return nil, err
 		}
 
-		// TODO error type
-		return nil, fmt.Errorf("request %v failed: %v: %s", request, response.Status, contents)
+		// Trim the trailing newline from HTTP error messages
+		if contents[len(contents)-1] == '\n' {
+			contents = contents[:len(contents)-1]
```
Comment (breerly) [inline transport/http/outbound.go]: can `len(contents)` ever be `0`?

File: transport/errors.go [hunk 2]
```diff
@@ -61,44 +77,106 @@ func AsHandlerError(service, procedure string, err error) error {
...
+func (e remoteBadRequestError) Error() string {
+	return e.Message
+}
+
+//////////////////////////////////////////////////////////////////////////////
 
 // UnexpectedError is a failure to process a request for an unexpected reason.
-type UnexpectedError struct {
-	isHandlerError
+type UnexpectedError interface {
```
Comment (breerly) [inline transport/errors.go]: Can you update this comment to add an easy-to-understand example of when users might run into `UnexpectedError`?

File: transport/http/handler_test.go [hunk 3]
```diff
@@ -56,41 +56,41 @@ func TestHandlerFailures(t *testing.T) {
...
 	}{
-		{&http.Request{Method: "GET"}, "not found"},
+		{&http.Request{Method: "GET"}, "404 page not found\n"},
 		{
 			&http.Request{
 				Method: "POST",
 				Header: headerCopyWithout(baseHeaders, CallerHeader),
 			},
-			"BadRequest: missing caller name",
+			"BadRequest: missing caller name\n",
```
Comment (breerly) [inline transport/http/handler_test.go]: Why add `\n` to these?

File: transport/tchannel/outbound.go [hunk 4]
```diff
@@ -135,3 +142,14 @@ func writeBody(body io.Reader, call *tchannel.OutboundCall) error {
...
 }
+
+func fromSystemError(err tchannel.SystemError) error {
+	switch err.Code() {
+	case tchannel.ErrCodeBadRequest:
+		return transport.RemoteBadRequestError(err.Message())
+	case tchannel.ErrCodeUnexpected:
+		return transport.RemoteUnexpectedError(err.Message())
+	default:
+		return err // TODO(abg): How to handle other error codes?
```
Comment (breerly) [inline transport/tchannel/outbound.go]: this seems like a pretty big matzo ball

**Comment by breerly** [comment]
Mostly nits. :shipit:

File: transport/errors.go [hunk 5]
```diff
@@ -61,44 +77,106 @@ func AsHandlerError(service, procedure string, err error) error {
...
 }
 
-func (e BadRequestError) Error() string {
+var _ BadRequestError = localBadRequestError{}
+var _ HandlerError = localBadRequestError{}
+
+// LocalBadRequestError wraps the given error into a BadRequestError.
+//
+// It represents a local failure while processing an invalid request.
+func LocalBadRequestError(err error) HandlerError {
```
Comment (abhinav) [inline transport/errors.go]: TODO: These functions should be inside an internal package so users cannot use them.

File: transport/http/outbound.go (hunk 1 above)
Comment (abhinav) [inline transport/http/outbound.go]: Oh, true. Good catch.

File: transport/errors.go (hunk 2 above)
Comment (abhinav) [inline transport/errors.go]: 👍

File: transport/http/handler_test.go (hunk 3 above)
Comment (abhinav) [inline transport/http/handler_test.go]: I changed the assert from `Contains` to `Equal`. The HTTP response always contains a trailing newline (for readability, especially if you're curl-ing from a terminal).

File: transport/tchannel/outbound.go (hunk 4 above)
Comment (abhinav) [inline transport/tchannel/outbound.go]: True. Created https://github.com/yarpc/yarpc/issues/54 to discuss this.

**Comment by abhinav** [comment]
Hold off on reviewing this. We discussed some changes to this.

**Comment by breerly** [comment]
Mainly, we might not need the distinction between local and remote errors if:

1. all errors returned from the handler are unexpected errors
2. application errors must be represented in a valid response (thrift exception, json obj, etc...)
3. bad request errors are used for malformed requests

**Comment by abhinav** [comment]
Ready for review

(a few questions were resolved offline)

File: transport/errors.go (hunk 5 above)
Comment (abhinav) [inline transport/errors.go]: Created https://github.com/yarpc/yarpc-go/issues/112

**Comment by abhinav** [comment]
Abandoning in favor of https://github.com/yarpc/yarpc-go/pull/114

**Pull request closed without merging**